File: leetcode/84/c++/solution.cpp
/*
    84. Largest Rectangle in Histogram
    Author: Miguel Angel Bermeo Ayerbe
*/
class Solution {
   public:
    int largestRectangleArea(vector<int>& heights) {
        std::stack<int> stack_heights;
        stack_heights.push(-1);
        int output = 0, height = 0, width = 1;
        for (size_t i = 0; i <= heights.size(); ++i) {
            while (stack_heights.top() >= 0 &&
                   (i == heights.size() ||
                    (i < heights.size() &&
                     heights[i] <= heights[stack_heights.top()]))) {
                height = heights[stack_heights.top()];
                stack_heights.pop();
                width = i - stack_heights.top() - 1;
                output = max(output, height * width);
            }
            if (i < heights.size()) {
                stack_heights.push(i);
            }
        }
        return output;
    }
};

File: leetcode/84/rust/solution.rs
/*
    84. Largest Rectangle in Histogram
    Author: Miguel Angel Bermeo Ayerbe
*/
impl Solution {
    pub fn largest_rectangle_area(heights: Vec<i32>) -> i32 {
        let mut stack: Vec<usize> = Vec::with_capacity(heights.len() + 1);
        stack.push(0);
        let mut max_area: i32 = 0;
        let mut stack_top: usize = 0;
        for i in 0..(heights.len() + 1) {
            stack_top = *stack.last().unwrap();
            while (stack_top != 0
                && (i == heights.len()
                    || (i < heights.len() && heights[i] < heights[stack_top - 1])))
            {
                let height = heights[stack_top - 1];
                stack.pop();
                stack_top = *stack.last().unwrap();
                let width = (i - stack_top) as i32;
                max_area = max_area.max(height * width);
            }
            stack.push(i + 1);
        }
        max_area
    }
}
